### PR TITLE
feat: add conversions  between snowflake id and  dateTime(dateTime64)

### LINKED ISSUE #27058: snowflake id (not to confuse with Snowflake service) time extract and convert functions
**Use case**
[snowflake id](https://en.wikipedia.org/wiki/Snowflake_ID) is a form of unique identifier used in distributed computing. The format was created by Twitter and it is widely used in many products such as discord api and etc. 

snowflake is a int64 number and generated based on timestamp, so it is comparable between two ids and it is sorting friendly,   therefore it is a good replacement to uuid  to be used as primary key in many occasions 

when it comes to clickhouse merge key engine,
- we can use snowflake to be the order key or primary key and extract time from it to be the partitioning key at the same time
- some time based queries can be applied on snowflack primary keys after converting time to snowflake and speed up the queries 

**Describe the solution you'd like**

I'd like to add a group of snowflake time extract and convert  functions to solve above questions, and I wonder whether the community will accept this feature.

### PR DESCRIPTION
**Pull request by jasine**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

add conversion functions  between snowflake id and  dateTime(dateTime64)  Close #27058 

Detailed description / Documentation draft:

following functions added

`snowflakeToDateTime`
`dateTimeToSnowflake`
`snowflakeToDateTime64`
`dateTime64ToSnowflake`

> By adding documentation, you'll allow users to try your new feature immediately, not when someone else will have time to document it later. Documentation is necessary for all features that affect user experience in any way. You can add brief documentation draft above, or add documentation right into your patch as Markdown files in [docs](https://github.com/ClickHouse/ClickHouse/tree/master/docs) folder.

> If you are doing this for the first time, it's recommended to read the lightweight [Contributing to ClickHouse Documentation](https://github.com/ClickHouse/ClickHouse/tree/master/docs/README.md) guide first.

> Information about CI checks: https://clickhouse.tech/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by kssenii** [comment]
Please, add tests (functional test in `tests/queries/0_stateless/` directory, [see link](https://clickhouse.tech/docs/en/development/tests/))

**Comment by jasine** [comment]
> Please, add tests (functional test in `tests/queries/0_stateless/` directory, [see link](https://clickhouse.tech/docs/en/development/tests/))

@kssenii tests and docs added, please review,  thanks.

File: src/Functions/dateTimeToSnowflake.cpp [hunk 1]
```diff
@@ -0,0 +1,14 @@
+#include <Functions/FunctionSnowflake.h>
+#include <Functions/FunctionFactory.h>
+
+namespace DB
+{
+
+void registerDateTimeToSnowflake(FunctionFactory & factory)
```
Comment (kssenii) [inline src/Functions/dateTimeToSnowflake.cpp:7]: May be put these to register functions in a single file?

**Review by kssenii — APPROVED** [review]

File: src/Functions/FunctionSnowflake.h [hunk 2]
```diff
@@ -54,7 +63,7 @@ class FunctionDateTimeToSnowflake : public IFunction
         const auto & source_data = typeid_cast<const ColumnUInt32 &>(col).getData();
         for (size_t i = 0; i < input_rows_count; ++i)
         {
-            result_data[i] = (int64_t(source_data[i])*1000-1288834974657)<<22;
+            result_data[i] = (Int64(source_data[i])*1000-snowflake_epoch)<<time_shift;
```
Comment (kssenii) [inline src/Functions/FunctionSnowflake.h:66]: can you please also add spaces between operators?

**Comment by jameingh** [comment]
@jasine Where is this code now?I can't find it in any branch.
image

**Comment by jasine** [comment]
@jameingh I force pushed when resolving conflict

**Pull request merged**